includes/data-port/class-sensei-data-port-job

```diff
@@ -0,0 +1,347 @@
+	 * Do question porting.
+	 *
+	 * @param array $question_state  Question sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                               'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_question_port( &$question_state );
+
+	/**
+	 * Do lesson porting.
+	 *
+	 * @param array $lesson_state  Lesson sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_lesson_port( &$lesson_state );
+
+	/**
+	 * Do course porting.
+	 *
+	 * @param array $course_state  Course sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                             'completed' when the port id completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_course_port( &$course_state );
+
+	/**
+	 * Serialize state to JSON.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			's' => $this->sub_jobs_state,
+			'l' => $this->logs,
+			'r' => $this->results,
+			'c' => $this->is_completed,
+			'p' => $this->percentage,
+		];
+	}
+
+	/**
+	 * Restore state from JSON.
+	 */
+	private function restore_from_json( $json_string ) {
+		$json_arr = json_decode( $json_string, true );
+
+		if ( ! $json_arr ) {
+			return;
+		}
+
+		$this->sub_jobs_state = $json_arr['s'];
+		$this->logs           = $json_arr['l'];
+		$this->results        = $json_arr['r'];
+		$this->is_completed   = $json_arr['c'];
+		$this->percentage     = $json_arr['p'];
+	}
+
+	/**
+	 * Add an entry to the logs.
+	 *
+	 * @param string $name      Name of the entity this log applies to.
+	 * @param string $message   Log message.
+	 * @param string $id        Id of the entity this log applies to.
+	 */
+	protected function add_log_entry( $name, $message, $id = '' ) {
+		$this->has_changed = true;
+
+		$entry = [
+			'name' => sanitize_text_field( $name ),
+			'msg'  => sanitize_text_field( $message ),
+		];
+
+		if ( ! empty( $id ) ) {
+			$entry['id'] = sanitize_text_field( $id );
+		}
+
+		$this->logs[] = $entry;
+	}
+
+	/**
+	 * Persist state to the db.
+	 */
+	public function persist() {
+		if ( $this->has_changed ) {
+			update_option( self::get_option_name(), wp_slash( wp_json_encode( $this ) ) );
+		}
+	}
+
+	/**
+	 * Run the job.
+	 */
+	public function run() {
+		if ( $this->is_completed || empty( $this->sub_jobs_state ) ) {
+			return;
+		}
+
+		$this->has_changed = true;
+		$this->percentage  = 0;
+		$subjob_percentage = 1 / count( $this->sub_jobs_state );
+
+		if ( ! empty($this->sub_jobs_state[ 'questions' ][ 'status' ]) ) {
+
+			if ( $this->sub_jobs_state[ 'questions' ][ 'status' ] !== 'completed' ) {
+				$this->percentage += $subjob_percentage * $this->do_question_port( $this->sub_jobs_state[ 'questions' ] );
```

Comment:
We could alternatively demand from subclasses to store the percentage in the state and calculate the percentage for the whole job in a separate job.